packages/ec-core/src/generator/mod.rs

```diff
@@ -1,83 +1,2 @@
-use anyhow::Context;
-use rand::{rngs::ThreadRng, seq::SliceRandom};
-
 pub mod collection;
-/// Trait to represent types that can generate (possibly random) instances
-/// of the generic type `T`. The implementing type provides whatever context
-/// is necessary to generate instances of `T`, such as probabilities or
-/// mechanisms for generating sub-types.
-pub trait Generator<T> {
-    /// # Errors
-    ///
-    /// This returns an `anyhow::Error` if the implementation of `generate`
-    /// returns some sort of error. An example would be choosing a random
-    /// item from a collection; this fails if the collection is empty.
-    fn generate(&self, rng: &mut ThreadRng) -> anyhow::Result<T>;
-}
-
-/// Implement `Generator` for any reference to a `Generator`.
-impl<'a, T, U> Generator<T> for &'a U
-where
-    U: Generator<T>,
-{
-    fn generate(&self, rng: &mut ThreadRng) -> anyhow::Result<T> {
-        (**self).generate(rng)
-    }
-}
-
-/// Implement `Generator` for any mutable reference to a `Generator`.
-impl<'a, T, U> Generator<T> for &'a mut U
-where
-    U: Generator<T>,
-{
-    fn generate(&self, rng: &mut ThreadRng) -> anyhow::Result<T> {
-        (**self).generate(rng)
-    }
-}
-
-/// Generate a random element from an array of options.
-///
-/// # Errors
-///
-/// This returns an `anyhow::Error` if the array of options
-/// is empty since we can't choose a random element from an
-/// empty array.
-impl<const N: usize, T> Generator<T> for [T; N]
```

Comment:
I see the value of `SliceCloning` below, but it seems like it still might be useful to be able to create `Distribution`s from arrays or vectors. We can create a `SliceCloning` from a reference to our array or vector (you do that in several places below), but there's definitely some machinery there.

Would it make sense to implement things like `impl<T> From<Vec<T>> for SliceCloning<T>` and similar for arrays? Or even `impl<T> From<IntoIterator<Item = T>> for SliceCloning<T>`?

Reply:
We would need some other kind of newtype since we have dangling references (which the compiler will not allow) if we try to convert a `Vec<T>` -> `SliceCloning<T>` since the SliceCloning doesn't own the actual data, and only references it. But that would probably just be another argument to move the GeneGenerator impl to be more general and allow all kinds of distributions.

Reply:
But that would probably be a reasonable addition, maybe along a `IntoDistribution` and / or `AsDistribution` Trait along the lines of 
```rs
trait IntoDistribution<T> {
    type Distribution: Distribution<T>;
    type Error;

    fn into_distribution(self) -> Result<Self::Distribution,Self::Error>;
}

trait AsDistribution<'a, T> {
    type Distribution: Distribution<T>;
    type Error;

    fn as_distribution(&'a self) -> Result<Self::Distribution, Self::Error>;
}

impl<T> IntoDistribution<T> for Vec<T> {
     type Distribution = VecDistr<T>;
     type Error = EmptySlice;
     
     fn into_distribution(self) -> Result<Self::Distribution,Self::Error> {
         VecDistr::new(self)
     }
}

impl<T> AsDistribution<T> for Vec<T>
where T: Clone {
    type Distribution = SliceCloning<'a, T>;
    type Error = EmptySlice;
     
     fn as_distribution(&'a self) -> Result<Self::Distribution,Self::Error> {
         SliceCloning::new(self)
     }
}

impl<'a, T> AsDistribution<&'a T> for Vec<T> {
    type Distribution = Slice<'a, T>;
    type Error = EmptySlice;
     
     fn as_distribution(&self) -> Result<Self::Distribution,Self::Error> {
         Slice::new(self).map_err(|_| EmptySlice)
     }
}
```

Reply:
We could then even take a `impl IntoDistribution<...>` in for example GeneGenerator, but that would need us handling the error there to, so I would probably just leave it as `impl Distribution<...>` and call that manually where we create it, that also allows us to just `?` it and doesn't require us handling that (unrelated) error in GeneGenerator itsself.

Reply:
I added these traits, maybe that also makes the above comments less relevant.